synchronized/lib/src/reentrant_lock.dart

```diff
@@ -11,11 +11,11 @@ class ReentrantLock implements Lock {
   final List<BasicLock> innerLocks = [BasicLock()];
 
   /// Inner level count.
-  int get innerLevel => (Zone.current[this] as int) ?? 0;
+  int get innerLevel => (Zone.current[this] as int?) ?? 0;
 
   @override
-  Future<T> synchronized<T>(FutureOr<T> Function() func,
-      {Duration timeout}) async {
+  Future<T?> synchronized<T>(FutureOr<T?> Function()? func,
```

Comment:
note: I don't think you need `T?` here - just `T` should be fine (`T` can represent a nullable type)

Reply:
Ah ok thanks

Reply:
Well after some testing I'm not sure...
```
// no lint error
Future<T?> action1<T>(List<T> list) async {}

// Lint error!
//  The body might complete normally, causing 'null' to be returned, but the return type is a potentially non-nullable type.
Future<T> action2<T>(List<T> list) async {}
```

That's puzzling indeed. Should I fill a bug?

Reply:
The reason I suggested `T` here is because you had `T?` as the return type of the function also - so that suggested to me that you are really just returning the value of `func` after invoking it.

Compare the usage from a users perspective of this function:

```
var result = await synchronized(() async => 1);
```

With a return type of `Future<T?>`, we get an inferred type of `int?` for `result` even though it definitely never returns null. So users have to add unnecessary casts or null checks.

Alternatively, with the return type of `Future<T>` you get an inferred type of `T` - or in this case `int`, which is a much more desirable type.

https://nullsafety.dartpad.dev/e71feb86c968725f868e6461a2732c20 is a gist demonstrating how the inference does work out if you do the cast.

Reply:
Note that you can work around the lint when returning `null`, by casting null to `T`. It is weird, but basically null is valid for all the valid uses of the function.

Users could make the as cast fail if they did this `syncronized<int>(null)`. But, that doesn't do anything useful anyways.

Reply:
> by casting null to `T`.

That was the trick indeed. Thanks for explanation and dartpad (really convenient to share shippets...). I replaced all `T?` by `T`. I don't think the migration tool did it, it had a blue square and using T? solved my return null issue.